### PR TITLE
feat: Add native fixed-array dot

### LINKED ISSUE #17456: Cannot `dot` two `Array(float64, N)` columns
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl

df = pl.DataFrame(
    {"a": [np.arange(3) for _ in range(2)], "b": [np.arange(3) + 10 for _ in range(2)]},
    schema={"a": pl.Array(pl.Float64, 3), "b": pl.Array(pl.Float64, 3)},
)

df.with_columns(pl.col("a").dot(pl.col("b")))
```

### Log output

```shell
Traceback (most recent call last):
  File "/Users/rben01/file.py", line 24, in <module>
    df.with_columns(pl.col("a").dot(pl.col("b")))
  File "/Users/rben01/Library/Caches/pypoetry/virtualenvs/base-wTklQcO3-py3.12/lib/python3.12/site-packages/polars/dataframe/frame.py", line 8763, in with_columns
    return self.lazy().with_columns(*exprs, **named_exprs).collect(_eager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/rben01/Library/Caches/pypoetry/virtualenvs/base-wTklQcO3-py3.12/lib/python3.12/site-packages/polars/lazyframe/frame.py", line 1942, in collect
    return wrap_df(ldf.collect(callback))
                   ^^^^^^^^^^^^^^^^^^^^^
polars.exceptions.InvalidOperationError: `sum` operation not supported for dtype `array[f64, 3]`
```

### Issue description

I guess that after computing the elementwise product, Polars fails to take the sum of the resulting array?

### Expected behavior

`0 * 10 + 1 * 11 + 2 * 12 = 35`

### Installed versions

<details>

```
--------Version info---------
Polars:               1.0.0
Index type:           UInt32
Platform:             macOS-14.5-arm64-arm-64bit
Python:               3.12.4 (main, Jun  6 2024, 18:26:44) [Clang 15.0.0 (clang-1500.3.9.4)]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
great_tables:         <not installed>
hvplot:               <not installed>
matplotlib:           3.9.1
nest_asyncio:         1.6.0
numpy:                2.0.0
openpyxl:             3.1.5
pandas:               2.2.2
pyarrow:              16.1.0
pydantic:             <not installed>
pyiceberg:            <not installed>
sqlalchemy:           2.0.31
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by 0guban0v**
Closes #17456 and benefits embedding pipelines doing similarity scoring or candidate reranking.

Related #24652 is tracked separately by #28614: vertical fixed-array aggregation Expr.sum() reduces Array rows into one Array, while arr.dot reduces each row into one scalar. 

Issue example uses `Expr.dot`, which reduces columns to one scalar. This PR adds `arr.dot` because requested operation is row-wise over each fixed-size array.

```python
  df.select(
      score=pl.col("embedding").arr.dot(query)
  )
```

API distinction:

```python
# Column reduction producing one scalar
pl.col("a").dot("b")

# Row-wise Array operation producing one scalar per row
pl.col("a").arr.dot("b")
```

Existing equivalent materializes a `rows × width` multiplication result before reducing:

```python
(pl.col("a") * pl.col("b")).arr.sum()
```

Native kernel fuses multiplication and reduction, so it does not materialize rows × width product. Fragmented inputs still rechunk, so peak memory can include contiguous copies of lhs and rhs in addition to scalar output.
Attached benchmark in comments.

Scope:
- matching `Float32` or `Float64` inner dtypes;
- equal widths;
- equal row counts or one-row broadcasting;
- outer null propagation;
- inner-null products ignored;
- fragmented inputs rechunk through existing behavior;
- raw Python sequences, one-dimensional NumPy arrays, and literal vectors are cast to the left Array dtype;
- non-literal expressions and Series still require exactly matching Array dtypes;

Limitations: 
- do not support Integer or mixed-float

**Update (2026-07-30)**: `arr.dot` now accepts Python sequences and one-dimensional NumPy arrays directly. These raw query vectors are cast after selector or wildcard expansion, when each left expression has concrete Array dtype. Explicit expressions and Series operands retain their dtype and must match the left operand.

AI disclosure:
OpenAI Codex drafted early benchmark code  and plots with my revision, fixed my grammar and phrasing.

### CODE REVIEW

**Comment by 0guban0v** [comment]
Possible follow-up, separate from this PR, is to fuse sibling `arr.dot` expressions that score same Array column against several constant query vectors.
 
``` python
df.select(
    score_a=pl.col("embedding").arr.dot(query_a),
    score_b=pl.col("embedding").arr.dot(query_b),
)
```

Today each expression traverses embedding independently. Planner could read each embedding array once, compute both scores in one multi-query kernel, and still return two separately named columns. Expressions that cannot be fused would continue using existing `arr.dot` execution.

Standalone Apple M4 experiment found this faster from two query vectors onward, but it has not been validated inside Polars and floating-point tolerance is unresolved. I don't want to mess with baseline.

**Comment by 0guban0v** [comment]
Addressed raw-query-vector API mismatch. `IntoExpr` includes Python lists and NumPy arrays, but these previously became List or primitive Series literals and failed fixed-size Array requirement. Basically, bump data science product fit.

**Comment by 0guban0v** [comment]
In-memory engine. Not much difference between Float 32 and Float 64. I will check rechunk impact on memory in the follow-up work.

<table>
  <tr>
    <td>m4-product-canary-trace-f32 </td>
    <td>m4-product-canary-peak-rss</td>
  </tr>
</table>

**Comment by 0guban0v** [comment]
@orlp , @ritchie46 , whenever you have time, this one is ready. I don't want to grow scope, it's reasonable with limitations and follow-up work. Happy to answer your questions.

**Comment by 0guban0v** [comment]
Verified current implementation explicitly on

```
  q.collect(engine="streaming")
  q.collect(engine="in-memory")
```
both produced identical results

**Comment by 0guban0v** [comment]
It’s ready for review, but it’s currently blocked by new, unrelated CI bugs. I have to convert it back to a draft due to my two-PR limit so I can prioritize CI fix. Feel free to review this PR anyway. 

@wence-  If you want to see additional measurements, let me know which.

**Comment by wence-** [comment]
> @wence- If you want to see additional measurements, let me know which.

I'm not sure I had thought to want measurements here at all? (perhaps here? https://github.com/rapidsai/cudf/pull/23583)

File: py-polars/src/polars/expr/array.py [hunk 1]
```diff
@@ -277,6 +279,96 @@ def sum(self) -> Expr:
...
+            Array expression or query vector to compute dot product with.
+            A Python sequence or one-dimensional NumPy array is treated as a
+            one-row query and cast to the data type of the input expression.
+            A one-row Series is broadcast and retains its data type.
+            Expression and Series inputs must have matching Array data types.
+
+        Notes
+        -----
+        Coordinates are paired strictly by position; no label alignment is
+        performed.
```
Comment (wence-) [inline py-polars/src/polars/expr/array.py]: it seems strange to talk about label alignment when polars has no labels, or any such concept IIUC.

File: py-polars/src/polars/expr/array.py [hunk 2]
```diff
@@ -277,6 +279,96 @@ def sum(self) -> Expr:
...
 
+    def dot(self, other: IntoExpr | Sequence[Any]) -> Expr:
+        """
+        Compute row-wise dot product with another Array expression.
+
+        Both inputs must contain equal-width arrays with matching ``Float32`` or
+        ``Float64`` inner dtypes.
+        Input with one row is broadcast against other input.
+        Products containing an inner null are ignored. An outer null row produces
+        a null.
```
Comment (wence-) [inline py-polars/src/polars/expr/array.py]: What does "Products containing an inner null are ignored" mean?

File: py-polars/src/polars/expr/array.py [hunk 3]
```diff
@@ -277,6 +279,96 @@ def sum(self) -> Expr:
...
+            Expression and Series inputs must have matching Array data types.
+
+        Notes
+        -----
+        Coordinates are paired strictly by position; no label alignment is
+        performed.
+
+        Coordinate pairs containing an inner null are skipped. If a non-null row
+        has no valid coordinate pairs, the result is ``0.0``. Similarity pipelines
+        should validate or fill inner nulls when zero must mean orthogonality.
```
Comment (wence-) [inline py-polars/src/polars/expr/array.py]: Where did coordinates come from?

File: py-polars/tests/unit/operations/namespaces/array/test_array.py [hunk 4]
```diff
@@ -80,6 +83,374 @@ def test_arr_sum(
...
+    )
+    expr = cs.by_dtype(pl.Array).arr.dot([10.0, 20.0])
+
+    binary = expr.meta.serialize(format="binary")
+    binary_expr = pl.Expr.deserialize(io.BytesIO(binary), format="binary")
+    assert_frame_equal(df.select(binary_expr), expected)
+
+    json = expr.meta.serialize(format="json")
+    json_expr = pl.Expr.deserialize(io.StringIO(json), format="json")
+    assert_frame_equal(df.select(json_expr), expected)
```
Comment (wence-) [inline py-polars/tests/unit/operations/namespaces/array/test_array.py]: question: I don't know the overall policy here, but my general impression is that tests don't specifically check that any given expression round-trips through the serializer. Indeed, the only other examples of deserialize in the test suite are specifically in test_serde.py

File: crates/polars-plan/src/plans/conversion/dsl_to_ir/functions.rs [hunk 5]
```diff
@@ -35,6 +35,18 @@ pub(super) fn convert_functions(
...
+                        let cast = ctx.arena.add(AExpr::Cast {
+                            expr: e[1].node(),
+                            dtype,
+                            options: CastOptions::Strict,
+                        });
+                        e[1].set_node(cast);
+                    }
+                    IA::Dot
+                },
                 A::ToList => IA::ToList,
```
Comment (wence-) [inline crates/polars-plan/src/plans/conversion/dsl_to_ir/functions.rs]: question: It seems odd to have this very special casting rule. Should would not rather apply the casting rules for `+` and `*` as seen in type_coercion.rs?

File: crates/polars-plan/src/plans/aexpr/function_expr/array.rs [hunk 6]
```diff
@@ -66,6 +67,47 @@ impl IRArrayFunction {
...
+                );
+                polars_ensure!(
+                    lhs_inner == rhs_inner,
+                    SchemaMismatch:
+                    "arr.dot requires matching inner dtypes, got {lhs_inner} and {rhs_inner}"
+                );
+                polars_ensure!(
+                    matches!(lhs_inner, DataType::Float32 | DataType::Float64),
+                    InvalidOperation:
+                    "arr.dot supports Float32 and Float64 arrays, got {}",
```
Comment (wence-) [inline crates/polars-plan/src/plans/aexpr/function_expr/array.rs]: It seems to me one could do this on any numeric type, what am I missing?

File: py-polars/src/polars/expr/array.py (hunk 2 above)
Comment (0guban0v) [reply py-polars/src/polars/expr/array.py]: test_arr_dot_nulls validates this.  If either input array is null for a row, final result for that row is null. I will make it clearer.

File: py-polars/src/polars/expr/array.py (hunk 1 above)
Comment (0guban0v) [reply py-polars/src/polars/expr/array.py]: Polars actually uses “label” for ML target columns, category names, and window/output labels, but Arrays do not have labeled elements, so “label alignment” is misleading wording here. I will fix it.

**Comment by ritchie46** [comment]
@0guban0v I exempted you from the PR limits. Thanks for your contributions.

File: py-polars/src/polars/expr/array.py (hunk 3 above)
Comment (0guban0v) [reply py-polars/src/polars/expr/array.py]: If we treat inputs to array.dot as vectors, “coordinates” is standard term for their elements. Here, coordinate pair means elements at the same position in two vectors. What's your recommendation here?

File: py-polars/tests/unit/operations/namespaces/array/test_array.py (hunk 4 above)
Comment (0guban0v) [reply py-polars/tests/unit/operations/namespaces/array/test_array.py]: **Edit: `test_arr_dot_query_vector_expr_serde`  is no longer required. Ignore below comment.**

`test_serde.py` is better location, I did not know about it. Thanks for sharing.

Here is example:
```python
import io

import polars as pl
import polars.selectors as cs

df = pl.DataFrame(
    {
        "f32": [[1.0, 2.0]],
        "f64": [[3.0, 4.0]],
    },
    schema={
        "f32": pl.Array(pl.Float32, 2),
        "f64": pl.Array(pl.Float64, 2),
    },
)

expr = cs.by_dtype(pl.Array).arr.dot([10.0, 20.0])

serialized = expr.meta.serialize(format="binary")
restored = pl.Expr.deserialize(io.BytesIO(serialized), format="binary")

print(df.select(restored))
```
This specific test adds value only for users who serialize and later deserialize the expression in mixed F32 and F64 setup. It is real edge-case.

File: crates/polars-plan/src/plans/aexpr/function_expr/array.rs (hunk 6 above)
Comment (0guban0v) [reply crates/polars-plan/src/plans/aexpr/function_expr/array.rs]: I planned to add integer support separately (see description and comments) because integer multiplication and accumulation require explicit overflow handling. Would you prefer this PR to support all numeric types?

Comment (wence-) [reply crates/polars-plan/src/plans/aexpr/function_expr/array.rs]: No, that's what I'm missing.

File: crates/polars-plan/src/plans/conversion/dsl_to_ir/functions.rs (hunk 5 above)
Comment (0guban0v) [reply crates/polars-plan/src/plans/conversion/dsl_to_ir/functions.rs]: Yes, array.dot should use standard supertype rules, and edge-case test for mixed F32/F64 is not needed. Current approach could quickly become spaghetti which runs counter to unification trend in this repo. I will fix this.

**Comment by 0guban0v** [comment]
There is issue with NumPy to scalar Array conversion. I will open all follow tickets to make pending scope clear.

**Comment by 0guban0v** [comment]
In-memory engine. No performance regression after refactoring. Follow up issues will cover further improvement.

<table>
  <tr>
    <td>m4-product-canary-trace-f32</td>
    <td>m4-product-canary-peak-rss
</td>
  </tr>
</table>

**Comment by 0guban0v** [comment]
generalized and cleaned up test cases

File: crates/polars-ops/src/chunked_array/array/dot.rs [hunk 7]
```diff
@@ -0,0 +1,133 @@
...
+    let lhs_inner_validity = lhs_values.validity();
+    let rhs_inner_validity = rhs_values.validity();
+    let width = lhs.width();
+    let lhs_broadcast = lhs.len() == 1 && output_len != 1;
+    let rhs_broadcast = rhs.len() == 1 && output_len != 1;
+
+    let mut output = Vec::with_capacity(output_len);
+    let mut output_validity = BitmapBuilder::with_capacity(output_len);
+
+    for output_idx in 0..output_len {
```
Comment (ritchie46) [inline crates/polars-ops/src/chunked_array/array/dot.rs:43]: There are a lot of bound checks here that we should be able to elide given that all the dimension are known up-front.

File: crates/polars-ops/src/chunked_array/array/dot.rs [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+    let (lhs_inner, lhs_width) = match lhs.dtype() {
+        DataType::Array(inner, width) => (inner.as_ref(), *width),
+        _ => unreachable!(),
+    };
+    let (rhs_inner, rhs_width) = match rhs.dtype() {
+        DataType::Array(inner, width) => (inner.as_ref(), *width),
+        _ => unreachable!(),
+    };
+
+    polars_ensure!(
```
Comment (ritchie46) [inline crates/polars-ops/src/chunked_array/array/dot.rs]: can these be asserts? This should already be checked during IR resolving. I see the checks in `get_field`.

File: py-polars/src/polars/expr/array.py [hunk 9]
```diff
@@ -277,6 +279,84 @@ def sum(self) -> Expr:
...
 
+    def dot(self, other: IntoExpr | Sequence[Any]) -> Expr:
+        """
+        Compute row-wise dot product with another Array expression.
+
+        Both inputs must contain equal-width arrays. Their inner data types are cast
+        to a common supertype, which must be ``Float32`` or ``Float64``.
+        An input with one row is broadcast against the other input.
+        If either input Array is null for a row, the result for that row is null.
+
```
Comment (ritchie46) [inline py-polars/src/polars/expr/array.py:292]: Can you add engine tags here?

File: py-polars/src/polars/series/array.py [hunk 10]
```diff
@@ -76,6 +76,53 @@ def sum(self) -> Series:
...
+        """
+        Compute row-wise dot product with another Array Series or query vector.
+
+        Both inputs must contain equal-width arrays. Their inner data types are cast
+        to a common supertype, which must be ``Float32`` or ``Float64``.
+        An input with one row is broadcast against the other input.
+        A Python sequence or one-dimensional NumPy array is treated as a one-row
+        Array query.
+        If either input Array is null for a row, the result for that row is null.
+
```
Comment (ritchie46) [inline py-polars/src/polars/series/array.py:89]: Can you add engine tags here?

File: py-polars/src/polars/expr/array.py (hunk 9 above)
Comment (0guban0v) [reply py-polars/src/polars/expr/array.py:292]: Done

File: py-polars/src/polars/series/array.py (hunk 10 above)
Comment (0guban0v) [reply py-polars/src/polars/series/array.py:89]: None of existing Series methods have engine tags. To avoid two implementations,  Series.arr.dot delegates to Expr.arr.dot through expr_dispatch. I guess we can treat comments as code metadata. Duplicating engine tags here can trigger drift in future. Are you ok not adding it here?

File: crates/polars-ops/src/chunked_array/array/dot.rs (hunk 8 above)
Comment (0guban0v) [reply crates/polars-ops/src/chunked_array/array/dot.rs]: Done

File: crates/polars-ops/src/chunked_array/array/dot.rs (hunk 7 above)
Comment (0guban0v) [reply crates/polars-ops/src/chunked_array/array/dot.rs:43]: Borrowed approach from one of your PRs. Ready for review.

File: py-polars/src/polars/series/array.py (hunk 10 above)
Comment (ritchie46) [reply py-polars/src/polars/series/array.py:89]: You are right, it should only be on expressions. NVM.

**Review by ritchie46 — APPROVED** [review]
